### PR TITLE
Rewrite multiple literal choice regex to multiple contains in rlike

### PR DESCRIPTION
**Pull request by thirtiseven**
Closes https://github.com/NVIDIA/spark-rapids/issues/10976
Fixes https://github.com/NVIDIA/spark-rapids/issues/10975

This pr rewrite multiple literal choice regex like `(aaa|bbb|ccc)` to multiple contains to speed up `rlike`.

Also fixed a bug that pattern like `^.*(aaa)` shouldn't be rewrite because `.` will not match line break characters.

**Performance results:**

Data: 50000000 string data generated with big data gen. 

Test case 1: or of 3 patterns with 15 length, repeat 20 times.
Test case 2: or of 20 patterns with 5 length, repeat 20 times.
Test case 3: or of 20 patterns with 50 length, repeat 2 times.

Time in ms:

| Test case | CPU    | Gluten | 24.08  | This PR | Speed up vs 24.08 |
|--------|--------|--------|--------|---------|----|
| 1   | 51,991 | 14,042 | 6,459  | 779     |  8.29x  |
| 2   | 163,348| 16,284 | 59,139 | 1,965   |  30.10x   |
| 3  | 8,196  | 2,643  | 42,046 | 684     |   61.47x  |

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -1126,6 +1127,33 @@ case class GpuRLike(left: Expression, right: Expression, pattern: String)
...
+  override def child: Expression = input
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType)
+
+  override def doColumnar(input: GpuColumnVector): ColumnVector = {
+    val inputLength = input.getRowCount.toInt
+    val accInit = withResource(Scalar.fromBool(false)) { falseScalar => 
+      ColumnVector.fromScalar(falseScalar, inputLength)
+    }
+    searchList.foldLeft(accInit) { (acc, search) =>
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1144]:
```suggestion
    val accInit = withResource(Scalar.fromString(searchList.head)) { searchScalar =>
      input.getBase.stringContains(searchScalar)
    }
    searchList.tail.foldLeft(accInit) { (acc, search) =>
```

Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: good suggestion, done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 2]
```diff
@@ -2091,6 +2092,30 @@ object RegexRewrite {
     }
   }
 
+  private def getMultipleContainsLiterals(ast: RegexAST): (Boolean, Seq[String]) = {
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala]: Yes, updated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 3]
```diff
@@ -2091,6 +2092,20 @@ object RegexRewrite {
     }
   }
 
+  private def getMultipleContainsLiterals(ast: RegexAST): Seq[String] = {
+    ast match {
+      case RegexGroup(_, term, _) => getMultipleContainsLiterals(term)
+      case RegexChoice(RegexSequence(parts), ls) if isLiteralString(parts) => {
+        getMultipleContainsLiterals(ls) match {
+          case ls if ls.isEmpty => Seq.empty
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:2100]: nit:
```suggestion
          case Seq() => Seq.empty
```

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by thirtiseven** [comment]
build

**Review by NVnavkumar — DISMISSED** [review]

**Review by gerashegalov — APPROVED** [review]

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala [hunk 4]
```diff
@@ -67,8 +67,27 @@ class RegularExpressionRewriteSuite extends AnyFunSuite {
...
+      ".*(abc|def|ghi).*",
+      "((abc)|(def))",
+      "(abc)|(def)",
+      "(火花|急流)"
+    )
+    val excepted = Seq(
+      MultipleContains(Seq("abc", "def")),
+      MultipleContains(Seq("abc", "def", "ghi")),
+      MultipleContains(Seq("abc", "def")),
+      NoOptimization,
```
Comment (HaoYang670) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala:89]: Just curious why does `((abc)|(def))` can be optimized but `(abc)|(def)` can not?

Comment (gerashegalov) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala:89]: An orthogonal high-level question: have we considered making these regex rewrites in libcudf such that other libcudf users benefit as well.

Comment (thirtiseven) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala:89]: > An orthogonal high-level question: have we considered making these regex rewrites in libcudf such that other libcudf users benefit as well.

Yes, a better way is to speed up the cudf regex engine itself, https://github.com/rapidsai/cudf/issues/15611 proposed to do similar optimization, but rewriting it from Spark side are some lower hanging fruits to take.

Comment (thirtiseven) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala]: > Just curious why does `((abc)|(def))` can be optimized but `(abc)|(def)` can not?

Good question, this is actually a bug:

In `GpuRLikeMeta` I called `matchSimplePattern` with `regexAst.children()` and assumed that most of the regexAst is a `RegexSequence`, but this is not always true. 

For example `(abc)|(def)` is a single `RegexChoice(RegexSequence(ListBuffer(RegexChar(a), RegexChar(b), RegexChar(c))), RegexSequence(ListBuffer(RegexChar(d), RegexChar(e), RegexChar(f))))`, so I pass the children of it and it failed to be optimized.

Fixed it.

**Comment by thirtiseven** [comment]
build

**Review by NVnavkumar — APPROVED** [review]
LGTM

**Pull request merged**